CMakeLists.txt

```diff
@@ -90,6 +90,63 @@ list(APPEND rcutils_sources
   include/rcutils/logging_macros.h)
 include_directories("${CMAKE_CURRENT_BINARY_DIR}/include")
 
+###############################################################################
+# Create version file in build/<package>/include/<package> folder which is already in the include list
+set(TMP_INCLUDE_DIR ${CMAKE_CURRENT_BINARY_DIR}/include/${PROJECT_NAME})
+set(RCUTILS_VERSION_FILE_NAME ${TMP_INCLUDE_DIR}/version.h)
+set(NEED_TO_CREATE_VERSION_FILE false)
+
+# retrieve verson information from <package>.xml file
+# call ament_package_xml() if it has not been called before
+if(NOT _AMENT_PACKAGE_NAME)
+  ament_package_xml()
+endif()
+# parse version information from the version string
+string(REGEX REPLACE "([0-9]+\).([0-9]+)\.([0-9]+)" "" dummy ${rcutils_VERSION})
+set(RCUTILS_VERSION_MAJOR ${CMAKE_MATCH_1})
+set(RCUTILS_VERSION_MINOR ${CMAKE_MATCH_2})
+set(RCUTILS_VERSION_PATCH ${CMAKE_MATCH_3})
+string(TIMESTAMP RCUTILS_VERSION_TIMESTAMP "%Y-%m-%dT%H:%M:%SZ" UTC)
+
+# sanity checks
+if("${RCUTILS_VERSION_MAJOR}" STREQUAL "")
+  message(FATAL_ERROR "RCUTILS_VERSION_MAJOR is empty or non-numeric, check <package>.XML")
+endif()
+if("${RCUTILS_VERSION_MINOR}" STREQUAL "")
+  message(FATAL_ERROR "RCUTILS_VERSION_MINOR is empty or non-numeric, check <package>.XML")
+endif()
+if("${RCUTILS_VERSION_PATCH}" STREQUAL "")
+  message(FATAL_ERROR "RCUTILS_VERSION_PATCH is empty or non-numeric, check <package>.XML")
+endif()
+
+# Check if the version file exist
+if(EXISTS "${RCUTILS_VERSION_FILE_NAME}")
+  # The file exists
+  # Check if it contains the same version
+  file(STRINGS ${RCUTILS_VERSION_FILE_NAME} VERSION_FILE_STRINGS REGEX "#define RCUTILS_VERSION_STR")
+  if(VERSION_FILE_STRINGS STREQUAL "")
+    set(NEED_TO_CREATE_VERSION_FILE true)
+  else()
+    string(REGEX REPLACE "^#define[ \t]+RCUTILS_VERSION_STR[ \t]+\"([0-9]+\.[0-9]+\.[0-9]+)\"" "" dummy ${VERSION_FILE_STRINGS})
+    if(NOT (${CMAKE_MATCH_1} STREQUAL ${rcutils_VERSION}))
+      # Create new file if file version != rcutils_VERSION
+      set(NEED_TO_CREATE_VERSION_FILE true)
+    endif()
+  endif()
+else()
+  # if the version file does not exist, create it
+  set(NEED_TO_CREATE_VERSION_FILE true)
+endif()
```

Comment:
This whole logic can go away if the generated file doesn't contain a timestamp and only depends on the version number coming from the manifest.

When the manifest changes it is already ensured that a CMake reconfiguration will happen again (see https://github.com/ament/ament_cmake/blob/be483fdcf62de7d89ea4107de2ea594804860ec4/ament_cmake_core/cmake/core/ament_package_xml.cmake#L73).

If the goal is to avoid touching the file in case the content wouldn't change a check could be added if the existing file contains the exact version string to skip the `configure_file` call.

Reply:
The goal is to avoid touching. The code check the following possible reasons to (re-)create the file:

- The version file does not exist
- The version file contains unparseable (eg corrupted) information
- The version file contains the wrong version